src/Interactive/Features/Interactive/Core/InteractiveHost.RemoteService.cs

```diff
@@ -49,35 +56,38 @@ internal RemoteService(InteractiveHost host, Process process, int processId, Ser
 
             internal void HookAutoRestartEvent()
             {
-                const int ProcessExitHooked = 1;
-                const int ProcessExitHandled = 2;
-
-                int processExitHandling = 0;
+                using(_disposeSemaphore.DisposableWait())
+                {
+                    // hook the event only once per process:
+                    if (_processExitHandling == 0)
+                    {
+                        Process.Exited += ProcessExitedHandler;
+                        _processExitHandling = ProcessExitHooked;
+                    }
+                }
+            }
 
-                async void localHandler(object _, EventArgs __)
+            private async void ProcessExitedHandler(object _, EventArgs __)
+            {
+                try
                 {
-                    try
+                    using (await _disposeSemaphore.DisposableWaitAsync().ConfigureAwait(false))
                     {
-                        if (Interlocked.Exchange(ref processExitHandling, ProcessExitHandled) == ProcessExitHooked)
+                        if (_processExitHandling == ProcessExitHooked)
                         {
-                            Process.Exited -= localHandler;
-
+                            Process.Exited -= ProcessExitedHandler;
                             if (!_disposing)
```

Comment:
>if (!_disposing) [](start = 28, length = 16)

I think this is not needed now. Once Dispose sets _processExitHandling to ProcessExitHandled we won't get here.